Spurious empty results in SqlAggregateTest

I observed it 3 times over the course of weeks. It was in these tests:
- `test_avg`
- `test_groupByHaving`

Maybe it's not even related to `SqlAggregateTest`. In every test it failed because the result was empty. The test passes hundreds of times when we try to reproduce it.

Logs from the `test_avg` test:
```
2020-11-03 10:03:49,033 [DEBUG] [test_avg] [c.h.j.s.i.JetSqlBackend]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Before logical opt:
LogicalAggregate(group=[{}], EXPR$0=[AVG($0)])
  LogicalProject(distance=[$1])
    LogicalTableScan(table=[[hazelcast, public, o_f74a0711_05d8_4b6d_8c2b_ec31c1ccee29[projects=[0, 1]]]])

2020-11-03 10:03:49,033 [DEBUG] [test_avg] [c.h.j.s.i.JetSqlBackend]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] After logical opt:
AggregateLogicalRel(group=[{}], EXPR$0=[AVG($0)])
  FullScanLogicalRel(table=[[hazelcast, public, o_f74a0711_05d8_4b6d_8c2b_ec31c1ccee29[projects=[1]]]])

2020-11-03 10:03:49,048 [DEBUG] [test_avg] [c.h.j.s.i.JetSqlBackend]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] After physical opt:
AggregateCombinePhysicalRel(group=[{}], EXPR$0=[AVG($0)])
  AggregateAccumulatePhysicalRel
    FullScanPhysicalRel(table=[[hazelcast, public, o_f74a0711_05d8_4b6d_8c2b_ec31c1ccee29[projects=[1]]]])

2020-11-03 10:03:49,064 [ INFO] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.JobCoordinationService]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Starting job 0537-951f-c690-0001 based on submit request
2020-11-03 10:03:49,086 [ INFO] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Didn't find any snapshot to restore for job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,086 [ INFO] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Start executing job '0537-951f-c690-0001', execution 0537-951f-c691-0001, execution graph in DOT format:
digraph DAG {
	"TestBatch[public.o_f74a0711_05d8_4b6d_8c2b_ec31c1ccee29]" [localParallelism=1];
	"Accumulate" [localParallelism=8];
	"Combine" [localParallelism=1];
	"ClientSink" [localParallelism=1];
	"TestBatch[public.o_f74a0711_05d8_4b6d_8c2b_ec31c1ccee29]" -> "Accumulate" [queueSize=1024];
	"Accumulate" -> "Combine" [label="distributed to [127.0.0.1]:5701-partitioned", queueSize=1024];
	"Combine" -> "ClientSink" [label="distributed to [127.0.0.1]:5701-partitioned", queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
2020-11-03 10:03:49,086 [DEBUG] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Building execution plan for job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,086 [DEBUG] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Built execution plans for job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,086 [DEBUG] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.o.InitExecutionOperation]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Initializing execution plan for job 0537-951f-c690-0001, execution 0537-951f-c691-0001 from [127.0.0.1]:5701
2020-11-03 10:03:49,086 [ INFO] [hz.romantic_nobel.cached.thread-4] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Execution plan for jobId=0537-951f-c690-0001, jobName='0537-951f-c690-0001', executionId=0537-951f-c691-0001 initialized
2020-11-03 10:03:49,086 [DEBUG] [hz.clever_nobel.generic-operation.thread-2] [c.h.j.i.o.InitExecutionOperation]: [127.0.0.1]:5702 [jet] [4.4-SNAPSHOT] Initializing execution plan for job 0537-951f-c690-0001, execution 0537-951f-c691-0001 from [127.0.0.1]:5701
2020-11-03 10:03:49,102 [ INFO] [hz.clever_nobel.generic-operation.thread-2] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5702 [jet] [4.4-SNAPSHOT] Execution plan for jobId=0537-951f-c690-0001, jobName='0537-951f-c690-0001', executionId=0537-951f-c691-0001 initialized
2020-11-03 10:03:49,102 [DEBUG] [hz.romantic_nobel.cached.thread-10] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Init of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 was successful
2020-11-03 10:03:49,102 [DEBUG] [hz.romantic_nobel.cached.thread-10] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Executing job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,102 [ INFO] [hz.romantic_nobel.cached.thread-10] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Start execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 from coordinator [127.0.0.1]:5701
2020-11-03 10:03:49,102 [ INFO] [hz.clever_nobel.generic-operation.thread-3] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5702 [jet] [4.4-SNAPSHOT] Start execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 from coordinator [127.0.0.1]:5701
2020-11-03 10:03:49,202 [DEBUG] [hz.romantic_nobel.jet.cooperative.thread-5] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 completed
2020-11-03 10:03:49,202 [DEBUG] [hz.clever_nobel.jet.cooperative.thread-0] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5702 [jet] [4.4-SNAPSHOT] Execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 completed
2020-11-03 10:03:49,218 [DEBUG] [ForkJoinPool.commonPool-worker-13] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 was successful
2020-11-03 10:03:49,218 [DEBUG] [hz.romantic_nobel.cached.thread-6] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Sending CompleteExecutionOperation for job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,218 [DEBUG] [hz.romantic_nobel.cached.thread-6] [c.h.j.i.o.CompleteExecutionOperation]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Completing execution 0537-951f-c691-0001 from caller [127.0.0.1]:5701, error=null
2020-11-03 10:03:49,218 [DEBUG] [hz.romantic_nobel.cached.thread-6] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Completed execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,218 [DEBUG] [hz.clever_nobel.generic-operation.thread-1] [c.h.j.i.o.CompleteExecutionOperation]: [127.0.0.1]:5702 [jet] [4.4-SNAPSHOT] Completing execution 0537-951f-c691-0001 from caller [127.0.0.1]:5701, error=null
2020-11-03 10:03:49,218 [DEBUG] [hz.clever_nobel.generic-operation.thread-1] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5702 [jet] [4.4-SNAPSHOT] Completed execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001
2020-11-03 10:03:49,218 [ INFO] [hz.romantic_nobel.cached.thread-6] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] Execution of job '0537-951f-c690-0001', execution 0537-951f-c691-0001 completed successfully
	Start time: 2020-11-03T10:03:49.064
	Duration: 154 ms
	To see additional job metrics enable JobConfig.storeMetricsAfterJobCompletion
2020-11-03 10:03:49,218 [DEBUG] [hz.romantic_nobel.cached.thread-6] [c.h.j.i.JobCoordinationService]: [127.0.0.1]:5701 [jet] [4.4-SNAPSHOT] job '0537-951f-c690-0001', execution 0537-951f-c691-0001 is completed
2020-11-03 10:03:49,218 [ INFO] [Time-limited test] [c.h.j.SimpleTestInClusterSupport]: Ditching 8 jobs in SimpleTestInClusterSupport.@After: [0537-951f-c688-0001, 0537-951f-c68a-0001, 0537-951f-c68c-0001, 0537-951f-c68e-0001, 0537-951f-c690-0001, 0537-951f-c682-0001, 0537-951f-c684-0001, 0537-951f-c686-0001]
2020-11-03 10:03:49,218 [ INFO] [Time-limited test] [c.h.j.SimpleTestInClusterSupport]: Destroying 1 distributed objects in SimpleTestInClusterSupport.@After: [hz:impl:replicatedMapService/__sql.catalog]
BuildInfo right after test_avg(com.hazelcast.jet.sql.SqlAggregateTest): BuildInfo{version='4.1-SNAPSHOT', build='20201023', buildNumber=20201023, revision=fac8b58, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='4.4-SNAPSHOT', build='20201103', revision='@git.commit.id.abbrev@'}}

No metrics recorded during the test

java.lang.AssertionError: 
Expecting:
  <[]>
to contain exactly in any order:
  <[Row{[3]}]>
but could not find the following elements:
  <[Row{[3]}]>


	at com.hazelcast.jet.sql.SqlTestSupport.assertRowsAnyOrder(SqlTestSupport.java:118)
	at com.hazelcast.jet.sql.SqlAggregateTest.test_avg(SqlAggregateTest.java:753)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)


```

--- a/hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetQueryResultProducer.java
+++ b/hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetQueryResultProducer.java
@@ -28,7 +28,7 @@ import com.hazelcast.sql.impl.row.Row;
 
 import java.util.NoSuchElementException;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static com.hazelcast.sql.impl.ResultIterator.HasNextResult.DONE;
 import static com.hazelcast.sql.impl.ResultIterator.HasNextResult.TIMEOUT;
@@ -40,11 +40,12 @@ public class JetQueryResultProducer implements QueryResultProducer {
 
     static final int QUEUE_CAPACITY = 4096;
 
+    private static final Exception NORMAL_COMPLETION = new NormalCompletionException();
+
     private final OneToOneConcurrentArrayQueue<Row> rows = new OneToOneConcurrentArrayQueue<>(QUEUE_CAPACITY);
-    private final AtomicBoolean done = new AtomicBoolean(false);
+    private final AtomicReference<Exception> done = new AtomicReference<>();
 
     private InternalIterator iterator;
-    private Exception exception;
 
     @Override
     public ResultIterator<Row> iterator() {
@@ -56,8 +57,8 @@ public class JetQueryResultProducer implements QueryResultProducer {
     }
 
     public void consume(Inbox inbox) {
-        if (done.get()) {
-            throw exception == null ? new IllegalStateException("Already done") : new RuntimeException(exception);
+        if (done.get() != null) {
+            throw new RuntimeException(done.get());
         }
         for (Object[] row; (row = (Object[]) inbox.peek()) != null && rows.offer(new HeapRow(row)); ) {
             inbox.remove();
@@ -65,14 +66,12 @@ public class JetQueryResultProducer implements QueryResultProducer {
     }
 
     public void done() {
-        done.set(true);
+        done.compareAndSet(null, NORMAL_COMPLETION);
     }
 
     @Override
     public void onError(QueryException error) {
-        if (done.compareAndSet(false, true)) {
-            exception = error;
-        }
+        done.compareAndSet(null, error);
     }
 
     private class InternalIterator implements ResultIterator<Row> {
@@ -89,27 +88,10 @@ public class JetQueryResultProducer implements QueryResultProducer {
 
         @Override
         public HasNextResult hasNext(long timeout, TimeUnit timeUnit) {
-            if (nextRow != null || (nextRow = rows.poll()) != null) {
-                return YES;
-            }
-
-            if (done.get()) {
-                if ((nextRow = rows.poll()) != null) {
-                    return YES;
-                } else {
-                    if (exception == null) {
-                        return DONE;
-                    } else {
-                        throw new RuntimeException("The Jet SQL job failed: " + exception.getMessage(), exception);
-                    }
-                }
-            }
-
-            if (timeout == 0) {
-                return TIMEOUT;
-            }
-
-            return hasNextWait(System.nanoTime() + timeUnit.toNanos(timeout));
+            return nextRow != null || (nextRow = rows.poll()) != null ? YES
+                    : isDone() ? DONE
+                    : timeout == 0 ? TIMEOUT
+                    : hasNextWait(System.nanoTime() + timeUnit.toNanos(timeout));
         }
 
         private HasNextResult hasNextWait(long endTimeNanos) {
@@ -118,22 +100,30 @@ public class JetQueryResultProducer implements QueryResultProducer {
                 if (nextRow != null || (nextRow = rows.poll()) != null) {
                     return YES;
                 }
-
-                if (done.get()) {
-                    if ((nextRow = rows.poll()) != null) {
-                        return YES;
-                    } else {
-                        if (exception == null) {
-                            return DONE;
-                        } else {
-                            throw new RuntimeException("The Jet SQL job failed: " + exception.getMessage(), exception);
-                        }
-                    }
+                if (isDone()) {
+                    return DONE;
                 }
-
                 idler.idle(++idleCount);
             } while (System.nanoTime() < endTimeNanos);
             return TIMEOUT;
+        }
+
+        /**
+         * Returns:<ul>
+         * <li>true, if done and rows is exhausted
+         * <li>false, if not done or rows are not exhausted
+         * <li>throws exception, if done with error
+         * </ul>
+         */
+        private boolean isDone() {
+            Exception exception = done.get();
+            if (exception != null) {
+                if (exception instanceof NormalCompletionException) {
+                    return rows.isEmpty();
+                }
+                throw new RuntimeException("The Jet SQL job failed: " + exception.getMessage(), exception);
+            }
+            return false;
         }
 
         @Override
@@ -146,6 +136,12 @@ public class JetQueryResultProducer implements QueryResultProducer {
             } finally {
                 nextRow = rows.poll();
             }
+        }
+    }
+
+    private static final class NormalCompletionException extends Exception {
+        NormalCompletionException() {
+            super("Done normally", null, false, false);
         }
     }
 }

--- a/hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetQueryResultProducer.java
+++ b/hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetQueryResultProducer.java
@@ -28,7 +28,7 @@ import com.hazelcast.sql.impl.row.Row;
 
 import java.util.NoSuchElementException;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static com.hazelcast.sql.impl.ResultIterator.HasNextResult.DONE;
 import static com.hazelcast.sql.impl.ResultIterator.HasNextResult.TIMEOUT;
@@ -40,36 +40,38 @@ public class JetQueryResultProducer implements QueryResultProducer {
 
     static final int QUEUE_CAPACITY = 4096;
 
-    private static final Exception NORMAL_COMPLETION = new NormalCompletionException();
+    private final OneToOneConcurrentArrayQueue<Row> rows = new OneToOneConcurrentArrayQueue<>(QUEUE_CAPACITY);
+    private final AtomicBoolean done = new AtomicBoolean(false);
 
-    private final OneToOneConcurrentArrayQueue<Row> queue = new OneToOneConcurrentArrayQueue<>(QUEUE_CAPACITY);
-    private final AtomicReference<Exception> done = new AtomicReference<>();
     private InternalIterator iterator;
+    private Exception exception;
 
     @Override
     public ResultIterator<Row> iterator() {
         if (iterator != null) {
-            throw new IllegalStateException("The iterator can be requested only once");
+            throw new IllegalStateException("Iterator can be requested only once");
         }
         iterator = new InternalIterator();
         return iterator;
     }
 
-    @Override
-    public void onError(QueryException error) {
-        done.compareAndSet(null, error);
+    public void consume(Inbox inbox) {
+        if (done.get()) {
+            throw exception == null ? new IllegalStateException("Already done") : new RuntimeException(exception);
+        }
+        for (Object[] row; (row = (Object[]) inbox.peek()) != null && rows.offer(new HeapRow(row)); ) {
+            inbox.remove();
+        }
     }
 
     public void done() {
-        done.compareAndSet(null, NORMAL_COMPLETION);
+        done.set(true);
     }
 
-    public void consume(Inbox inbox) {
-        if (done.get() != null) {
-            throw new RuntimeException(done.get());
-        }
-        for (Object[] r; (r = (Object[]) inbox.peek()) != null && queue.offer(new HeapRow(r)); ) {
-            inbox.remove();
+    @Override
+    public void onError(QueryException error) {
+        if (done.compareAndSet(false, true)) {
+            exception = error;
         }
     }
 
@@ -81,16 +83,57 @@ public class JetQueryResultProducer implements QueryResultProducer {
         private Row nextRow;
 
         @Override
-        public HasNextResult hasNext(long timeout, TimeUnit timeUnit) {
-            return nextRow != null || (nextRow = queue.poll()) != null ? YES
-                    : isDone() ? DONE
-                    : timeout == 0 ? TIMEOUT
-                    : hasNextWait(System.nanoTime() + timeUnit.toNanos(timeout));
+        public boolean hasNext() {
+            return hasNextWait(Long.MAX_VALUE) == YES;
         }
 
         @Override
-        public boolean hasNext() {
-            return hasNextWait(Long.MAX_VALUE) == YES;
+        public HasNextResult hasNext(long timeout, TimeUnit timeUnit) {
+            if (nextRow != null || (nextRow = rows.poll()) != null) {
+                return YES;
+            }
+
+            if (done.get()) {
+                if ((nextRow = rows.poll()) != null) {
+                    return YES;
+                } else {
+                    if (exception == null) {
+                        return DONE;
+                    } else {
+                        throw new RuntimeException("The Jet SQL job failed: " + exception.getMessage(), exception);
+                    }
+                }
+            }
+
+            if (timeout == 0) {
+                return TIMEOUT;
+            }
+
+            return hasNextWait(System.nanoTime() + timeUnit.toNanos(timeout));
+        }
+
+        private HasNextResult hasNextWait(long endTimeNanos) {
+            long idleCount = 0;
+            do {
+                if (nextRow != null || (nextRow = rows.poll()) != null) {
+                    return YES;
+                }
+
+                if (done.get()) {
+                    if ((nextRow = rows.poll()) != null) {
+                        return YES;
+                    } else {
+                        if (exception == null) {
+                            return DONE;
+                        } else {
+                            throw new RuntimeException("The Jet SQL job failed: " + exception.getMessage(), exception);
+                        }
+                    }
+                }
+
+                idler.idle(++idleCount);
+            } while (System.nanoTime() < endTimeNanos);
+            return TIMEOUT;
         }
 
         @Override
@@ -101,46 +144,8 @@ public class JetQueryResultProducer implements QueryResultProducer {
             try {
                 return nextRow;
             } finally {
-                nextRow = queue.poll();
+                nextRow = rows.poll();
             }
-        }
-
-        private HasNextResult hasNextWait(long endTimeNanos) {
-            long idleCount = 0;
-            do {
-                if (nextRow != null || (nextRow = queue.poll()) != null) {
-                    return YES;
-                }
-                if (isDone()) {
-                    return DONE;
-                }
-                idler.idle(++idleCount);
-            } while (System.nanoTime() < endTimeNanos);
-            return TIMEOUT;
-        }
-
-        /**
-         * Returns:<ul>
-         *     <li>true, if done
-         *     <li>false, if not done
-         *     <li>throws exception, if done with error
-         * </ul>
-         */
-        private boolean isDone() {
-            Exception doneExc = done.get();
-            if (doneExc != null) {
-                if (doneExc instanceof NormalCompletionException) {
-                    return true;
-                }
-                throw new RuntimeException("The Jet SQL job failed: " + doneExc.getMessage(), doneExc);
-            }
-            return false;
-        }
-    }
-
-    private static final class NormalCompletionException extends Exception {
-        NormalCompletionException() {
-            super("done normally");
         }
     }
 }

--- a/hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetQueryResultProducer.java
+++ b/hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetQueryResultProducer.java
@@ -56,6 +56,15 @@ public class JetQueryResultProducer implements QueryResultProducer {
         return iterator;
     }
 
+    @Override
+    public void onError(QueryException error) {
+        done.compareAndSet(null, error);
+    }
+
+    public void done() {
+        done.compareAndSet(null, NORMAL_COMPLETION);
+    }
+
     public void consume(Inbox inbox) {
         if (done.get() != null) {
             throw new RuntimeException(done.get());
@@ -63,15 +72,6 @@ public class JetQueryResultProducer implements QueryResultProducer {
         for (Object[] row; (row = (Object[]) inbox.peek()) != null && rows.offer(new HeapRow(row)); ) {
             inbox.remove();
         }
-    }
-
-    public void done() {
-        done.compareAndSet(null, NORMAL_COMPLETION);
-    }
-
-    @Override
-    public void onError(QueryException error) {
-        done.compareAndSet(null, error);
     }
 
     private class InternalIterator implements ResultIterator<Row> {
@@ -82,16 +82,28 @@ public class JetQueryResultProducer implements QueryResultProducer {
         private Row nextRow;
 
         @Override
-        public boolean hasNext() {
-            return hasNextWait(Long.MAX_VALUE) == YES;
-        }
-
-        @Override
         public HasNextResult hasNext(long timeout, TimeUnit timeUnit) {
             return nextRow != null || (nextRow = rows.poll()) != null ? YES
                     : isDone() ? DONE
                     : timeout == 0 ? TIMEOUT
                     : hasNextWait(System.nanoTime() + timeUnit.toNanos(timeout));
+        }
+
+        @Override
+        public boolean hasNext() {
+            return hasNextWait(Long.MAX_VALUE) == YES;
+        }
+
+        @Override
+        public Row next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            try {
+                return nextRow;
+            } finally {
+                nextRow = rows.poll();
+            }
         }
 
         private HasNextResult hasNextWait(long endTimeNanos) {
@@ -110,7 +122,7 @@ public class JetQueryResultProducer implements QueryResultProducer {
 
         /**
          * Returns:<ul>
-         * <li>true, if done and rows is exhausted
+         * <li>true, if done and rows are exhausted
          * <li>false, if not done or rows are not exhausted
          * <li>throws exception, if done with error
          * </ul>
@@ -119,28 +131,19 @@ public class JetQueryResultProducer implements QueryResultProducer {
             Exception exception = done.get();
             if (exception != null) {
                 if (exception instanceof NormalCompletionException) {
+                    // finish the rows first
                     return rows.isEmpty();
                 }
                 throw new RuntimeException("The Jet SQL job failed: " + exception.getMessage(), exception);
             }
             return false;
         }
-
-        @Override
-        public Row next() {
-            if (!hasNext()) {
-                throw new NoSuchElementException();
-            }
-            try {
-                return nextRow;
-            } finally {
-                nextRow = rows.poll();
-            }
-        }
     }
 
     private static final class NormalCompletionException extends Exception {
         NormalCompletionException() {
+            // Use writableStackTrace = false, the exception is not created at a place where it's thrown,
+            // it's better if it has no stack trace then.
             super("Done normally", null, false, false);
         }
     }

--- a/hazelcast-jet-sql/src/test/java/com/hazelcast/jet/sql/impl/JetQueryResultProducerTest.java
+++ b/hazelcast-jet-sql/src/test/java/com/hazelcast/jet/sql/impl/JetQueryResultProducerTest.java
@@ -33,15 +33,13 @@ import static com.hazelcast.sql.impl.ResultIterator.HasNextResult.YES;
 import static java.util.concurrent.TimeUnit.DAYS;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
 
 public class JetQueryResultProducerTest extends JetTestSupport {
 
-    private final JetQueryResultProducer p = new JetQueryResultProducer();
-    private final ResultIterator<Row> iterator = p.iterator();
+    private final JetQueryResultProducer producer = new JetQueryResultProducer();
+    private final ResultIterator<Row> iterator = producer.iterator();
     private final ArrayDequeInbox inbox = new ArrayDequeInbox(new ProgressTracker());
 
     @Test
@@ -49,12 +47,12 @@ public class JetQueryResultProducerTest extends JetTestSupport {
         Semaphore semaphore = new Semaphore(0);
         Future<?> future = spawn(() -> {
             try {
-                assertEquals(TIMEOUT, iterator.hasNext(0, SECONDS));
+                assertThat(iterator.hasNext(0, SECONDS)).isEqualTo(TIMEOUT);
                 semaphore.release();
-                assertTrue(iterator.hasNext());
+                assertThat(iterator.hasNext()).isTrue();
                 assertInstanceOf(Row.class, iterator.next());
                 semaphore.release();
-                assertFalse(iterator.hasNext());
+                assertThat(iterator.hasNext()).isFalse();
                 assertThatThrownBy(iterator::next)
                         .isInstanceOf(NoSuchElementException.class);
                 semaphore.release();
@@ -69,19 +67,19 @@ public class JetQueryResultProducerTest extends JetTestSupport {
 
         // check that the thread is blocked in `hasNext` - that it did not release the 2nd permit
         sleepMillis(50);
-        assertEquals(0, semaphore.availablePermits());
+        assertThat(semaphore.availablePermits()).isZero();
 
         inbox.queue().add(new Object[0]);
-        p.consume(inbox);
+        producer.consume(inbox);
 
         // 2nd permit - the row returned from the iterator
         semaphore.acquire();
 
         // check that the thread is blocked in `hasNext` - that it did not release the 2nd permit
         sleepMillis(50);
-        assertEquals(0, semaphore.availablePermits());
+        assertThat(semaphore.availablePermits()).isZero();
 
-        p.done();
+        producer.done();
 
         assertTrueEventually(future::isDone, 5);
         semaphore.acquire();
@@ -92,48 +90,63 @@ public class JetQueryResultProducerTest extends JetTestSupport {
 
     @Test
     public void when_done_then_remainingItemsIterated() {
-        inbox.queue().add(new Object[] {1});
-        inbox.queue().add(new Object[] {2});
-        p.consume(inbox);
-        p.done();
+        inbox.queue().add(new Object[]{1});
+        inbox.queue().add(new Object[]{2});
+        producer.consume(inbox);
+        producer.done();
 
-        assertTrue(iterator.hasNext());
-        assertEquals(1, (int) iterator.next().get(0));
-        assertTrue(iterator.hasNext());
-        assertEquals(2, (int) iterator.next().get(0));
-        assertFalse(iterator.hasNext());
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat((int) iterator.next().get(0)).isEqualTo(1);
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat((int) iterator.next().get(0)).isEqualTo(2);
+        assertThat(iterator).isExhausted();
     }
 
     @Test
-    public void when_doneWhileWaiting_then_throw_async() {
-        assertEquals(TIMEOUT, iterator.hasNext(0, SECONDS));
-        p.onError(QueryException.error("mock error"));
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    public void when_doneWithError_then_remainingItemsIterated() {
+        inbox.queue().add(new Object[]{1});
+        inbox.queue().add(new Object[]{2});
+        producer.consume(inbox);
+        producer.onError(QueryException.error("mock error"));
+
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat((int) iterator.next().get(0)).isEqualTo(1);
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat((int) iterator.next().get(0)).isEqualTo(2);
+        assertThatThrownBy(iterator::hasNext).hasMessageContaining("mock error");
+    }
+
+    @Test
+    public void when_doneWithErrorWhileWaiting_then_throw_async() {
+        assertThat(iterator.hasNext(0, SECONDS)).isEqualTo(TIMEOUT);
+        producer.onError(QueryException.error("mock error"));
         assertThatThrownBy(() -> iterator.hasNext(0, SECONDS))
                 .hasMessageContaining("mock error");
     }
 
     @Test
-    public void when_doneWhileWaiting_then_throw_sync() throws Exception {
+    public void when_doneWithErrorWhileWaiting_then_throw_sync() throws Exception {
         Future<?> future = spawn(() -> {
             assertThatThrownBy(() -> iterator.hasNext(1, DAYS))
                     .hasMessageContaining("mock error");
         });
         sleepMillis(50); // sleep so that the thread starts blocking in `hasNext`
-        p.onError(QueryException.error("mock error"));
+        producer.onError(QueryException.error("mock error"));
         future.get();
     }
 
     @Test
     public void when_nextItemWhileWaiting_then_hasNextReturns() throws Exception {
         Future<?> future = spawn(() -> {
-            assertEquals(YES, iterator.hasNext(1, DAYS));
-            assertEquals(42, (int) iterator.next().get(0));
+            assertThat(iterator.hasNext(1, DAYS)).isEqualTo(YES);
+            assertThat((int) iterator.next().get(0)).isEqualTo(42);
         });
         sleepMillis(50); // sleep so that the thread starts blocking in `hasNext`
 
         inbox.queue().add(new Object[]{42});
-        p.consume(inbox);
-        assertEquals(0, inbox.size());
+        producer.consume(inbox);
+        assertThat(inbox).isEmpty();
         future.get();
     }
 
@@ -142,38 +155,39 @@ public class JetQueryResultProducerTest extends JetTestSupport {
         long start = System.nanoTime();
         iterator.hasNext(500, MILLISECONDS);
         long elapsed = MILLISECONDS.toNanos(System.nanoTime() - start);
-        assertTrue("elapsed=" + elapsed, elapsed >= 500);
+        assertThat(elapsed >= 500).isTrue();
     }
 
     @Test
     public void when_iteratorRequestedTheSecondTime_then_fail() {
-        assertThatThrownBy(() -> p.iterator())
+        assertThatThrownBy(producer::iterator)
                 .hasMessageContaining("can be requested only once");
     }
 
     @Test
     public void when_onErrorAfterDone_then_ignored() {
-        p.onError(QueryException.error("error1"));
-        p.onError(QueryException.error("error2"));
+        producer.done();
+        producer.onError(QueryException.error("error"));
 
-        assertThatThrownBy(() -> iterator.hasNext())
+        assertThat(iterator.hasNext()).isFalse();
+    }
+
+    @Test
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    public void when_onErrorCalledTwice_then_secondIgnored() {
+        producer.onError(QueryException.error("error1"));
+        producer.onError(QueryException.error("error2"));
+
+        assertThatThrownBy(iterator::hasNext)
                 .hasMessageContaining("error1");
     }
 
     @Test
-    public void when_onErrorCalledTwice_then_secondIgnored() {
-        p.done();
-        p.onError(QueryException.error("error2"));
-
-        assertFalse(iterator.hasNext());
-    }
-
-    @Test
     public void when_doneCalledTwice_then_secondIgnored() {
-        p.done();
-        p.done();
+        producer.done();
+        producer.done();
 
-        assertFalse(iterator.hasNext());
+        assertThat(iterator.hasNext()).isFalse();
     }
 
     @Test
@@ -182,7 +196,7 @@ public class JetQueryResultProducerTest extends JetTestSupport {
         for (int i = 0; i < JetQueryResultProducer.QUEUE_CAPACITY + numExcessItems; i++) {
             inbox.queue().add(new Object[0]);
         }
-        p.consume(inbox);
-        assertEquals(2, inbox.size());
+        producer.consume(inbox);
+        assertThat(inbox).hasSize(2);
     }
 }

--- a/hazelcast-jet-sql/src/test/java/com/hazelcast/jet/sql/impl/JetQueryResultProducerTest.java
+++ b/hazelcast-jet-sql/src/test/java/com/hazelcast/jet/sql/impl/JetQueryResultProducerTest.java
@@ -103,21 +103,6 @@ public class JetQueryResultProducerTest extends JetTestSupport {
     }
 
     @Test
-    @SuppressWarnings("ResultOfMethodCallIgnored")
-    public void when_doneWithError_then_remainingItemsIterated() {
-        inbox.queue().add(new Object[]{1});
-        inbox.queue().add(new Object[]{2});
-        producer.consume(inbox);
-        producer.onError(QueryException.error("mock error"));
-
-        assertThat(iterator.hasNext()).isTrue();
-        assertThat((int) iterator.next().get(0)).isEqualTo(1);
-        assertThat(iterator.hasNext()).isTrue();
-        assertThat((int) iterator.next().get(0)).isEqualTo(2);
-        assertThatThrownBy(iterator::hasNext).hasMessageContaining("mock error");
-    }
-
-    @Test
     public void when_doneWithErrorWhileWaiting_then_throw_async() {
         assertThat(iterator.hasNext(0, SECONDS)).isEqualTo(TIMEOUT);
         producer.onError(QueryException.error("mock error"));

